### PR TITLE
Remove ILCompiler package reference

### PR DESCRIPTION
**Pull request by agocke**
It doesn't look like we need the newest version of the package anymore. This configuration (using the package instead of the SDK version) can be dangerous because it causes a split between the version built against vs. the version AOT'd against.

### CODE REVIEW

**Review by sbomer — APPROVED** [review]
LGTM, thank you!

**Comment by MichalStrehovsky** [comment]
Fixes #106740?

File: src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj [hunk 1]
```diff
@@ -18,11 +18,6 @@
     <SuppressGenerateILCompilerExplicitPackageReferenceWarning>true</SuppressGenerateILCompilerExplicitPackageReferenceWarning>
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(UseNativeAotForComponents)' == 'true'">
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj:21]: We may want to keep this around and condition it on `MicrosoftDotNetILCompilerVersion` being set. It is likely that we will need to take newer ILC that what's in SDK at some point in future again to take a critical fix temporarily.

Comment (jkotas) [reply src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj:21]: (We are on similar plan with Roslyn: We use Roslyn from SDK in steady state, but we may use a newer one temporarily.)

**Comment by jkotas** [comment]
Is this going to expose us to #105441 in more places until we update the SDK to RC1? (We should not block this PR on this - just making sure that it does not come as a surprise.)

File: src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj [hunk 2]
```diff
@@ -15,10 +15,10 @@
     <PublishReadyToRun Condition="'$(UseNativeAotForComponents)' != 'true'">true</PublishReadyToRun>
     <PublishSingleFile Condition="'$(UseNativeAotForComponents)' != 'true'">true</PublishSingleFile>
     <PublishTrimmed Condition="'$(UseNativeAotForComponents)' != 'true'">true</PublishTrimmed>
-    <SuppressGenerateILCompilerExplicitPackageReferenceWarning>true</SuppressGenerateILCompilerExplicitPackageReferenceWarning>
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(UseNativeAotForComponents)' == 'true'">
+  <ItemGroup Condition="'$(UseNativeAotForComponents)' == 'true' and '(MicrosoftDotNetILCompilerVersion)' != ''">
+    <SuppressGenerateILCompilerExplicitPackageReferenceWarning>true</SuppressGenerateILCompilerExplicitPackageReferenceWarning>
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj:21]: SuppressGenerateILCompilerExplicitPackageReferenceWarning is a property, but this is ItemGroup

File: src/native/managed/native-library.props [hunk 3]
```diff
@@ -34,9 +33,4 @@
     <!-- passed by compile-native.proj to set - -gcc-toolchain=$(ROOTFS_DIR)/usr -->
     <CustomLinkerArg Condition="'$(CustomLinkerArgToolchainArg)' != ''" Include="$(CustomLinkerArgToolchainArg)" />
   </ItemGroup>
-
-  <ItemGroup>
```
Comment (jkotas) [inline src/native/managed/native-library.props:38]: Conditionally include when MicrosoftDotNetILCompilerVersion is non-empty like in the other place?

Comment (agocke) [reply src/native/managed/native-library.props:38]: Lemme see if I can find a place to refactor this where it will get shared with evertyhing.

Comment (agocke) [reply src/native/managed/native-library.props:38]: OK, put everything in liveBuilds.targets. It's not the ideal location, but it's the best place I could find that's imported by everything that needs the override.

File: eng/Versions.props [hunk 4]
```diff
@@ -210,7 +210,6 @@
     <CompilerPlatformTestingMicrosoftVisualStudioCompositionVersion>17.0.46</CompilerPlatformTestingMicrosoftVisualStudioCompositionVersion>
     <!-- Docs -->
     <MicrosoftPrivateIntellisenseVersion>8.0.0-preview-20230918.1</MicrosoftPrivateIntellisenseVersion>
-    <!-- Mono Cecil -->
     <MicrosoftDotNetCecilVersion>0.11.5-alpha.24460.1</MicrosoftDotNetCecilVersion>
```
Comment (jkotas) [inline eng/Versions.props:213]:
```suggestion
    <!-- Mono Cecil -->
    <MicrosoftDotNetCecilVersion>0.11.5-alpha.24460.1</MicrosoftDotNetCecilVersion>
```

File: eng/Versions.props [hunk 5]
```diff
@@ -210,7 +210,6 @@
...
     <!-- ILCompiler -->
     <MicrosoftDotNetILCompilerVersion>9.0.0-rc.1.24410.5</MicrosoftDotNetILCompilerVersion>
```
Comment (jkotas) [inline eng/Versions.props:215]:
```suggestion
```

**Review by jkotas — APPROVED** [review]

**Review by elinor-fung — APPROVED** [review]

**Review by MichalStrehovsky — APPROVED** [review]

**Comment by MichalStrehovsky** [comment]
> Is this going to expose us to #105441 in more places until we update the SDK to RC1? (We should not block this PR on this - just making sure that it does not come as a surprise.)

I have update to RC1 lined up at #107673 so that should mitigate the issue if we can merge it shortly after this one.

**Comment by agocke** [comment]
/backport to release/9.0

I think this puts us in a better place for servicing than where we currently are

**Pull request merged**